test/e2e/xfn_test.go

```diff
@@ -240,3 +240,155 @@ func TestXfnRunnerImagePull(t *testing.T) {
+				if err := config.Client().Resources().List(ctx, nodes); err != nil {
+					t.Fatal("cannot list nodes", err)
+				}
+				if len(nodes.Items) == 0 {
+					t.Fatalf("no nodes in the cluster")
+				}
+
+				var addr string
+				for _, a := range nodes.Items[0].Status.Addresses {
+					if a.Type == corev1.NodeInternalIP {
+						addr = a.Address
+						break
+					}
+				}
+				if addr == "" {
+					t.Fatalf("no nodes with private address")
+				}
+
+				srcRef, err := name.ParseReference("crossplane-e2e/fn-tmp-writer:latest")
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				src, err := daemon.Image(srcRef)
+				if err != nil {
+					t.Fatal(err)
+				}
+				
+				err = wait.For(func() (done bool, err error) {
+					fmt.Println(fmt.Sprintf("%s:32000/fn-tmp-writer:latest", addr))
+					err = crane.Push(src, fmt.Sprintf("%s:32000/fn-tmp-writer:latest", addr), crane.Insecure)
+					if err != nil {
+						return false, nil //nolint:nilerr // we want to retry and to throw error
+					}
+					return true, nil
+				}, wait.WithTimeout(1*time.Minute))
+				if err != nil {
+					t.Fatal("copying image to registry not successful", err)
+				}
+				return ctx
+			}).
+			WithSetup("CrossplaneDeployedWithFunctionsEnabled", funcs.AllOf(
+				funcs.AsFeaturesFunc(funcs.HelmUpgrade(
+					HelmOptions(
+						helm.WithArgs(
+							"--set args={--debug,--enable-composition-functions}",
+							"--set xfn.enabled=true",
+							"--set xfn.args={--debug}",
+							"--set registryCaBundleConfig.name=reg-ca",
+							"--set registryCaBundleConfig.key=domain.crt",
+							"--set xfn.resources.requests.cpu=100m",
+							"--set xfn.resources.limits.cpu=100m",
+						),
+						helm.WithWait())...)),
+				funcs.ReadyToTestWithin(1*time.Minute, namespace),
+			)).
+			WithSetup("ProviderNopDeployed", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "prerequisites/provider.yaml"),
+				funcs.ApplyResources(FieldManager, manifests, "prerequisites/definition.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "prerequisites/provider.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "prerequisites/definition.yaml"),
+				funcs.ResourcesHaveConditionWithin(1*time.Minute, manifests, "prerequisites/definition.yaml", v1.WatchingComposite()),
+			)).
+			Assess("CompositionWithFunctionIsCreated", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "composition.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "composition.yaml"),
+			)).
+			Assess("ClaimIsCreated", funcs.AllOf(
+				funcs.ApplyResources(FieldManager, manifests, "claim.yaml"),
+				funcs.ResourcesCreatedWithin(30*time.Second, manifests, "claim.yaml"),
+			)).
+			Assess("ClaimBecomesAvailable", funcs.ResourcesHaveConditionWithin(5*time.Minute, manifests, "claim.yaml", xpv1.Available())).
+			Assess("ManagedResourcesProcessedByFunction", func(ctx context.Context, t *testing.T, config *envconf.Config) context.Context {
```

Comment:
Could this be rewritten to use `funcs.ResourceHasFieldValueWithin`? 🤔 Same goes for the `TestXfnRunnerImagePull` test case above probably

Reply:
`funcs.ResourceHasFieldValueWithin` requires a `k8s.Object` which is not convenient to use with claims/composites. I'll refactor to create a new helper to use.